From ff78ae0e2ea92adb869ba3c8b9095fc94bf04e52 Mon Sep 17 00:00:00 2001
From: Joachim Nilsson <troglobit@gmail.com>
Date: Wed, 5 Feb 2020 05:54:23 +0100
Subject: [PATCH] Fix #16: Regression in ifTable for non-Ethernet
 interfaces, Linux

This is an ugly workaround for Linux ppp and tun interfaces.  Ideally we
should instead factor out the IP address code introduced in 1.5 into the
get_ipinfo() function instead and maintain get_netinfo() only for iface
stats used by ifTable and ifXtable.

Signed-off-by: Joachim Nilsson <troglobit@gmail.com>
Signed-off-by: Alexander Sverdlin <alexander.sverdlin@gmail.com>
---
 linux.c      | 14 +++++++++-----
 mini-snmpd.h |  1 +
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/linux.c b/linux.c
index 8626a6a..c54011f 100644
--- a/linux.c
+++ b/linux.c
@@ -220,11 +220,13 @@ void get_netinfo(netinfo_t *netinfo)
 		if (i == -1)
 			continue;
 
+		logit(LOG_ERR, 0, "Interface %s is interesting! :-)", ifa->ifa_name);
 		switch (ifa->ifa_addr->sa_family) {
 		case AF_INET:
 			if (!ifa->ifa_addr || !ifa->ifa_netmask)
 				continue;
 
+			logit(LOG_ERR, 0, "Interface %s has a addr + netmask ...", ifa->ifa_name);
 			addr = (struct sockaddr_in *)ifa->ifa_addr;
 			mask = (struct sockaddr_in *)ifa->ifa_netmask;
 			if (addr) {
@@ -243,7 +245,12 @@ void get_netinfo(netinfo_t *netinfo)
 			/* XXX: Not supported yet */
 			break;
 
-		case AF_PACKET:
+		default:
+			break;
+		}
+
+		if (!netinfo->stats[i]) {
+			logit(LOG_ERR, 0, "Interface %s has an stats ...", ifa->ifa_name);
 			if (ifa->ifa_flags & IFF_POINTOPOINT)
 				netinfo->if_type[i] = 23; /* ppp(23) */
 			else if (ifa->ifa_flags & IFF_LOOPBACK)
@@ -283,10 +290,7 @@ void get_netinfo(netinfo_t *netinfo)
 
 			/* XXX: Need better tracking on Linux, c.f. FreeBSD ... */
 			netinfo->lastchange[1] = get_process_uptime();
-			break;
-
-		default:
-			break;
+			netinfo->stats[i] = 1;
 		}
 	}
 
diff --git a/mini-snmpd.h b/mini-snmpd.h
index 0f51762..ce8ec00 100644
--- a/mini-snmpd.h
+++ b/mini-snmpd.h
@@ -221,6 +221,7 @@ typedef struct netinfo_s {
 	unsigned int ifindex[MAX_NR_INTERFACES];
 	unsigned int status[MAX_NR_INTERFACES];
 	unsigned int lastchange[MAX_NR_INTERFACES];
+	unsigned int stats[MAX_NR_INTERFACES];		/* Sentinel for backends */
 	long long rx_bytes[MAX_NR_INTERFACES];
 	long long rx_mc_packets[MAX_NR_INTERFACES];
 	long long rx_bc_packets[MAX_NR_INTERFACES];
-- 
2.23.0

